content/post/2020-01-02-cran-error/index.Rmd

```diff
@@ -0,0 +1,106 @@
+---
+slug: cran-error
+title: "From Shock to Competence: How Not to Panic When You Receive E-mail from CRAN about Failed Checks"
+authors:
+  - Julia Romanowska
+date: "`r Sys.Date()`"
+tags:
+- package development
+output: 
+  html_document:
+    keep_md: true
+---
+
+> This post was contributed by [Julia Romanowska](https://jrom.bitbucket.io/homepage/), Researcher at the University of Bergen, Norway. Thank you, Julia!
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(fig.path = "", comment = "")
+
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+```
+
+<!--
+  - how we got to know about it
+  - raising an issue on bitbucket(?)
+  - installing docker and checking the docs
+  - searching for R and docker
+  - found r-hub and rocker
+  - running r-hub - problems
+  - figure out where the problem is (docker images didn't have access to the net)
+  - creating a script that runs everything manually, using the image provided by rhub
+-->
+
+I'm involved in development of the [Haplin](https://folk.uib.no/gjessing/genetics/software/haplin) R package, which enables fast genetic association analyses (very useful for those involved in genetic epidemiology research). We are a team of scientists that have various background, from genetics, through bioinformatics and statistics. Here's a short text about our latest "adventure" with CRAN and how it taught us some useful stuff.
+
+## OMG, CRAN wrote to me!
+
+There will come the time when you check your mailbox and see this dreaded message about a check that your beloved CRAN package failed. Even though the deadline for submitting a fix is quite reasonable, you might imagine yourself sitting long into the night over your laptop just to find out where is this one comma that was the source of the error. However, I need to say we were positively surprised over the level of details of feedback we got from CRAN concerning the error in _our_ package.
+
+In our case, the error was related to the changes that will be introduced in the next R version and that include throwing an error when the ``if`` statement resolves to a vector of values instead of one boolean value. In our code, we used to test for specific type of input data using ``class( obj ) == "myClass"``. This is sometimes resolved to a vector, which thus leads to an error in the `r-devel` CRAN tests. [This blog post](https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html) explains it in more details.
```

Comment:
Out of curiosity, in the end do you think it was worth running the checks locally as opposed to doing more guessing and sending edited versions of the package to the R-hub package builder?

Reply:
Yes, I always like to have control at least in the beginning, when I'm not really sure whether my corrections would result in less or more errors ;)